### PR TITLE
Feature "Plugins as Event Listeners"

### PR DESCRIPTION
**Pull request by fneumann**
https://www.ilias.de/docu/goto.php?target=wiki_1357_Plugins_as_Event_Listeners

### CODE REVIEW

File: Services/Component/classes/class.ilPlugin.php [hunk 1]
```diff
@@ -857,7 +857,13 @@ protected function beforeUninstall()
 				$ilDB->manipulate("DELETE FROM lng_modules".
 					" WHERE module = ".$ilDB->quote($prefix, "text"));
 			}
-			
+
+			// clear the event listening
+			include_once("Services/Component/classes/class.ilPluginReader.php");
+			$reader = new ilPluginReader($this->getDirectory() . '/plugin.xml',
+				$this->getComponentType(), $this->getComponentName(), $this->getSlotId(), $this->getPluginName());
+			$reader->clearEvents();
```
Comment (klees) [inline Services/Component/classes/class.ilPlugin.php]: Please move this to a function `clearEventListening`, that would spare the comment above and is good for general fitness.

File: Services/Component/classes/class.ilPlugin.php [hunk 2]
```diff
@@ -915,6 +921,13 @@ function update()
 		$ilCtrl->insertCtrlCalls("ilobjcomponentsettingsgui", ilPlugin::getConfigureClassName($this->getPluginName()),
 			$this->getPrefix());
 
+		// read the plugin.xml to get the event listening
+		include_once("Services/Component/classes/class.ilPluginReader.php");
+		$reader = new ilPluginReader($this->getDirectory() . '/plugin.xml',
+			$this->getComponentType(), $this->getComponentName(), $this->getSlotId(), $this->getPluginName());
+		$reader->clearEvents();
+		$reader->startParsing();
```
Comment (klees) [inline Services/Component/classes/class.ilPlugin.php]: Please move this to a function `readEventListening`, that would spare the comment above and is good for general fitness.

File: Services/Component/classes/class.ilPlugin.php [hunk 3]
```diff
@@ -915,6 +921,13 @@ function update()
 		$ilCtrl->insertCtrlCalls("ilobjcomponentsettingsgui", ilPlugin::getConfigureClassName($this->getPluginName()),
 			$this->getPrefix());
 
+		// read the plugin.xml to get the event listening
+		include_once("Services/Component/classes/class.ilPluginReader.php");
```
Comment (klees) [inline Services/Component/classes/class.ilPlugin.php]: We do not need this anymore, we have a classmap for autoloading now.

File: Services/EventHandling/classes/class.ilAppEventHandler.php [hunk 4]
```diff
@@ -111,19 +111,43 @@ function raise($a_component, $a_event, $a_parameter = "")
...
+				}
+				else
+				{
+					$class = 'il'.substr($listener,$last_slash + 1).'AppEventListener';
+					$file = "./".$listener."/classes/class.".$class.".php";
+
+					// detemine class and file
+					#$comp = explode("/", $listener);
+					#$class = "il".$comp[1]."AppEventListener";
+					#$file = "./".$listener."/classes/class.".$class.".php";
```
Comment (klees) [inline Services/EventHandling/classes/class.ilAppEventHandler.php]: Please remove this.

**Review by klees — COMMENTED** [review]
LGTM, minor comments regarding some cleanup.

Thx!

**Comment by klees** [comment]
I think @smeyer-ilias or @alex40724 want to have a look as well, or who is responsible for the event-stuff atm?

**Review by alex40724 — APPROVED** [review]
Looks good. Thanks.

File: Services/Component/classes/class.ilPlugin.php [hunk 5]
```diff
@@ -857,7 +857,13 @@ protected function beforeUninstall()
 				$ilDB->manipulate("DELETE FROM lng_modules".
 					" WHERE module = ".$ilDB->quote($prefix, "text"));
 			}
-			
+
+			// clear the event listening
+			include_once("Services/Component/classes/class.ilPluginReader.php");
```
Comment (chfsx) [inline Services/Component/classes/class.ilPlugin.php]: The include can be ommitted, just run a `composer dumpautoload` in libs/composer and the class will be part of the classmap and will be found in the autoloader

File: Services/Component/classes/class.ilPlugin.php (hunk 3 above)
Comment (chfsx) [reply Services/Component/classes/class.ilPlugin.php]: The include can be ommitted, just run a `composer dumpautoload` in libs/composer and the class will be part of the classmap and will be found in the autoloader

**Review by chfsx — APPROVED** [review]
LGTM in general, thanks!

**Pull request merged**